examples/src/main/java/com/squareup/moshi/recipes/EnvelopeJsonAdapterFactory.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonQualifier;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class EnvelopeJsonAdapterFactory implements JsonAdapter.Factory {
+  private static final class Train {
+    String name;
+  }
+
+  public static void main(String[] args) {
+    Enveloped annotation = new Enveloped() {
```

Comment:
We just need a better way to refer to an annotation. Implementing it isn’t good, especially since it has particular requirements for equals() and hashCode().